builtin/v13/migration/market.go

```diff
@@ -0,0 +1,590 @@
+		}
+	} else {
+		providerSectorsRoot, newStateArrayRoot, err = m.migrateProviderSectorsAndStatesFromScratch(ctx, store, in, states)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to migrate provider sectors (all): %w", err)
+		}
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealStatesInKey(in.Address), states); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous inRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealProposalsInKey(in.Address), proposals); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous inRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealStatesOutKey(in.Address), newStateArrayRoot); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous outRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevProviderSectorsOutKey(in.Address), providerSectorsRoot); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous outRoot to cache: %w", err)
+	}
+
+	return providerSectorsRoot, newStateArrayRoot, nil
+}
+
+func (m *marketMigrator) migrateProviderSectorsAndStatesWithDiff(ctx context.Context, store cbor.IpldStore, prevInStatesCid, prevOutStatesCid, prevOutProviderSectorsCid, inStatesCid, proposals cid.Cid) (cid.Cid, cid.Cid, error) {
+	diffs, err := amt.Diff(ctx, store, store, prevInStatesCid, inStatesCid, amt.UseTreeBitWidth(market12.StatesAmtBitwidth))
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to diff old and new deal state AMTs: %w", err)
+	}
+
+	ctxStore := adt.WrapStore(ctx, store)
+
+	prevOutStates, err := adt.AsArray(ctxStore, prevOutStatesCid, market13.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load prevOutStates array: %w", err)
+	}
+
+	prevOutProviderSectors, err := adt.AsMap(ctxStore, prevOutProviderSectorsCid, market13.ProviderSectorsHamtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load prevOutProviderSectors map: %w", err)
+	}
+
+	proposalsArr, err := adt.AsArray(ctxStore, proposals, market12.ProposalsAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load proposals map: %w", err)
+	}
+
+	// in-memory maps with changesets to be applied to prevOutProviderSectors
+	providerSectorsMem := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{}        // added
+	providerSectorsMemRemoved := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{} // removed
+
+	addProviderSectorEntry := func(deal abi.DealID, newState *market13.DealState) (abi.SectorNumber, error) {
+		sid, ok := m.providerSectors.dealToSector[deal]
+		if !ok {
+			return 0, xerrors.Errorf("deal %d not found in providerSectors", deal) // todo is this normal and possible??
```

Comment:
I don't think so, but I'm not sure.

Reply:
Isn't this the case annotated "// FIP: if such a sector cannot be found, assert that the deal's end epoch has passed and use sector ID 0" in the FromScratch case (in which case yes)?

 Adding the FIP spec as comments in this version too would be helpful. Though this code isn't setting the deal state sector number.

Reply:
This function is called inside of a `oldState.SlashEpoch != -1` so I believe that this isn't this case, there the sector number will correctly be set to zero

```
		case amt.Add: // both in case of chain going forward and chain reorg this is just new deals
			var oldState market12.DealState
			if err := oldState.UnmarshalCBOR(bytes.NewReader(change.After.Raw)); err != nil {
				return cid.Cid{}, cid.Cid{}, xerrors.Errorf("failed to unmarshal old state: %w", err)
			}

			var newState market13.DealState
			newState.SlashEpoch = oldState.SlashEpoch
			newState.LastUpdatedEpoch = oldState.LastUpdatedEpoch
			newState.SectorStartEpoch = oldState.SectorStartEpoch
			newState.SectorNumber = 0 // terminated / not found (?)

			if oldState.SlashEpoch == -1 {
				si, err := addProviderSectorEntry(deal)
```